File: generate_pdf.py
import re
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.units import mm, inch
from reportlab.lib import colors
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle, Preformatted
from reportlab.lib.enums import TA_LEFT, TA_CENTER

def clean_emoji(text):
    """Remove emojis and special unicode characters for PDF compatibility"""
    emoji_pattern = re.compile("["
        u"\U0001F600-\U0001F64F"
        u"\U0001F300-\U0001F5FF"
        u"\U0001F680-\U0001F6FF"
        u"\U0001F1E0-\U0001F1FF"
        u"\U00002702-\U000027B0"
        u"\U000024C2-\U0001F251"
        u"\U0001f926-\U0001f937"
        u"\U00010000-\U0010ffff"
        u"\u2640-\u2642"
        u"\u2600-\u2B55"
        u"\u200d"
        u"\u23cf"
        u"\u23e9"
        u"\u231a"
        u"\ufe0f"
        u"\u3030"
        "]+", flags=re.UNICODE)
    return emoji_pattern.sub(r'', text).strip()

def md_to_pdf(md_file, pdf_file):
    # Read markdown file
    with open(md_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Create PDF
    doc = SimpleDocTemplate(
        pdf_file,
        pagesize=A4,
        leftMargin=20*mm,
        rightMargin=20*mm,
        topMargin=20*mm,
        bottomMargin=20*mm
    )
    
    styles = getSampleStyleSheet()
    
    # Custom styles
    title_style = ParagraphStyle(
        'CustomTitle',
        parent=styles['Title'],
        fontSize=24,
        spaceAfter=20,
        alignment=TA_CENTER
    )
    
    h1_style = ParagraphStyle(
        'H1',
        parent=styles['Heading1'],
        fontSize=18,
        spaceBefore=20,
        spaceAfter=10,
        textColor=colors.HexColor('#1a5276')
    )
    
    h2_style = ParagraphStyle(
        'H2',
        parent=styles['Heading2'],
        fontSize=14,
        spaceBefore=15,
        spaceAfter=8,
        textColor=colors.HexColor('#2e86c1')
    )
    
    h3_style = ParagraphStyle(
        'H3',
        parent=styles['Heading3'],
        fontSize=12,
        spaceBefore=10,
        spaceAfter=6,
        textColor=colors.HexColor('#5499c7')
    )
    
    body_style = ParagraphStyle(
        'CustomBody',
        parent=styles['BodyText'],
        fontSize=10,
        leading=14,
        spaceAfter=6
    )
    
    code_style = ParagraphStyle(
        'Code',
        parent=styles['Code'],
        fontSize=8,
        leading=10,
        fontName='Courier',
        backColor=colors.HexColor('#f4f4f4'),
        borderWidth=1,
        borderColor=colors.HexColor('#dddddd'),
        borderPadding=5
    )
    
    story = []
    
    # Process markdown line by line
    lines = content.split('\n')
    in_code_block = False
    code_content = []
    
    for line in lines:
        # Handle code blocks
        if line.strip().startswith('```'):
            if in_code_block:
                # End of code block
                code_text = '\n'.join(code_content)
                if code_text.strip():
                    story.append(Spacer(1, 5))
                    story.append(Preformatted(code_text, code_style))
                    story.append(Spacer(1, 5))
                code_content = []
                in_code_block = False
            else:
                # Start of code block
                in_code_block = True
            continue
        
        if in_code_block:
            code_content.append(line)
            continue
        
        # Skip empty lines
        if not line.strip():
            story.append(Spacer(1, 6))
            continue
        
        # Clean emojis
        clean_line = clean_emoji(line)
        if not clean_line:
            continue
        
        # Headers
        if clean_line.startswith('# ') and not clean_line.startswith('## '):
            text = clean_emoji(clean_line[2:])
            story.append(Paragraph(text, title_style))
        elif clean_line.startswith('## '):
            text = clean_emoji(clean_line[3:])
            story.append(Paragraph(text, h1_style))
        elif clean_line.startswith('### '):
            text = clean_emoji(clean_line[4:])
            story.append(Paragraph(text, h2_style))
        elif clean_line.startswith('#### '):
            text = clean_emoji(clean_line[5:])
            story.append(Paragraph(text, h3_style))
        # Horizontal rules
        elif clean_line.strip() in ['---', '***', '___']:
            story.append(Spacer(1, 10))
        # Lists
        elif clean_line.startswith('- ') or clean_line.startswith('* '):
            text = clean_emoji(clean_line[2:])
            story.append(Paragraph(f"  {text}", body_style))
        elif re.match(r'^\d+\.', clean_line):
            text = clean_emoji(clean_line)
            story.append(Paragraph(text, body_style))
        # Code inline (backticks)
        elif '`' in clean_line:
            # Replace backticks with font tags
            text = re.sub(r'`([^`]+)`', r'<font face="Courier" size="9">\1</font>', clean_line)
            text = clean_emoji(text)
            story.append(Paragraph(text, body_style))
        # Regular text
        else:
            # Handle bold and italic
            text = clean_line
            text = re.sub(r'\*\*([^*]+)\*\*', r'<b>\1</b>', text)
            text = re.sub(r'\*([^*]+)\*', r'<i>\1</i>', text)
            story.append(Paragraph(text, body_style))
    
    # Build PDF
    doc.build(story)
    print(f"PDF gerado com sucesso: {pdf_file}")

if __name__ == "__main__":
    md_file = r"C:\Users\07548262523\Documents\stock\CONTINUIDADE.md"
    pdf_file = r"C:\Users\07548262523\Documents\stock\CONTINUIDADE.pdf"
    md_to_pdf(md_file, pdf_file)

File: generate_pdf.js
const PDFDocument = require('pdfkit');
const fs = require('fs');
const path = require('path');

// Read markdown file
const mdFile = path.join(__dirname, 'CONTINUIDADE.md');
const content = fs.readFileSync(mdFile, 'utf8');

// Create PDF
const pdfFile = path.join(__dirname, 'CONTINUIDADE.pdf');
const doc = new PDFDocument({
    size: 'A4',
    margin: 50,
    info: {
        Title: 'Projeto STOCK - Continuidade',
        Author: 'MimoCode',
        Subject: 'Documentação do Projeto'
    }
});

const stream = fs.createWriteStream(pdfFile);
doc.pipe(stream);

// Helper function to clean emojis
function cleanEmoji(text) {
    return text.replace(/[\u{1F600}-\u{1F64F}\u{1F300}-\u{1F5FF}\u{1F680}-\u{1F6FF}\u{1F1E0}-\u{1F1FF}\u{2702}-\u{27B0}\u{24C2}-\u{1F251}\u{1f926}-\u{1f937}\u{10000}-\u{10ffff}\u{2640}-\u{2642}\u{2600}-\u{2B55}\u{200d}\u{23cf}\u{23e9}\u{231a}\ufe0f\u{3030}]/gu, '').trim();
}

// Process content
const lines = content.split('\n');
let inCodeBlock = false;
let codeContent = [];

// Colors
const colors = {
    title: '#1a5276',
    h1: '#2e86c1',
    h2: '#5499c7',
    body: '#333333',
    code: '#666666'
};

// Font sizes
const fontSizes = {
    title: 24,
    h1: 18,
    h2: 14,
    h3: 12,
    body: 10,
    code: 9
};

let yPos = 50;

for (const line of lines) {
    // Handle code blocks
    if (line.trim().startsWith('```')) {
        if (inCodeBlock) {
            // End of code block - render code
            if (codeContent.length > 0) {
                yPos += 10;
                doc.font('Courier')
                    .fontSize(fontSizes.code)
                    .fillColor(colors.code);
                
                for (const codeLine of codeContent) {
                    if (yPos > 750) {
                        doc.addPage();
                        yPos = 50;
                    }
                    doc.text(codeLine, 60, yPos);
                    yPos += 12;
                }
                yPos += 10;
            }
            codeContent = [];
            inCodeBlock = false;
        } else {
            inCodeBlock = true;
        }
        continue;
    }
    
    if (inCodeBlock) {
        codeContent.push(line);
        continue;
    }
    
    // Skip empty lines
    if (!line.trim()) {
        yPos += 8;
        continue;
    }
    
    // Clean emojis
    const cleanLine = cleanEmoji(line);
    if (!cleanLine) continue;
    
    // Check if we need a new page
    if (yPos > 750) {
        doc.addPage();
        yPos = 50;
    }
    
    // Headers
    if (cleanLine.startsWith('# ') && !cleanLine.startsWith('## ')) {
        yPos += 20;
        doc.font('Helvetica-Bold')
            .fontSize(fontSizes.title)
            .fillColor(colors.title)
            .text(cleanLine.substring(2), 50, yPos);
        yPos += 30;
    } else if (cleanLine.startsWith('## ')) {
        yPos += 15;
        doc.font('Helvetica-Bold')
            .fontSize(fontSizes.h1)
            .fillColor(colors.h1)
            .text(cleanLine.substring(3), 50, yPos);
        yPos += 25;
    } else if (cleanLine.startsWith('### ')) {
        yPos += 10;
        doc.font('Helvetica-Bold')
            .fontSize(fontSizes.h2)
            .fillColor(colors.h2)
            .text(cleanLine.substring(4), 50, yPos);
        yPos += 20;
    } else if (cleanLine.startsWith('#### ')) {
        yPos += 8;
        doc.font('Helvetica-Bold')
            .fontSize(fontSizes.h3)
            .fillColor(colors.body)
            .text(cleanLine.substring(5), 50, yPos);
        yPos += 18;
    }
    // Horizontal rules
    else if (['---', '***', '___'].includes(cleanLine.trim())) {
        yPos += 10;
        doc.moveTo(50, yPos)
            .lineTo(545, yPos)
            .stroke('#cccccc');
        yPos += 10;
    }
    // Lists
    else if (cleanLine.startsWith('- ') || cleanLine.startsWith('* ')) {
        doc.font('Helvetica')
            .fontSize(fontSizes.body)
            .fillColor(colors.body)
            .text('  • ' + cleanLine.substring(2), 60, yPos);
        yPos += 14;
    }
    // Numbered lists
    else if (/^\d+\./.test(cleanLine)) {
        doc.font('Helvetica')
            .fontSize(fontSizes.body)
            .fillColor(colors.body)
            .text('  ' + cleanLine, 60, yPos);
        yPos += 14;
    }
    // Code inline
    else if (cleanLine.includes('`')) {
        const parts = cleanLine.split('`');
        let xPos = 60;
        
        for (let i = 0; i < parts.length; i++) {
            if (i % 2 === 0) {
                // Normal text
                if (parts[i]) {
                    doc.font('Helvetica')
                        .fontSize(fontSizes.body)
                        .fillColor(colors.body)
                        .text(parts[i], xPos, yPos, { continued: true });
                }
            } else {
                // Code text
                if (parts[i]) {
                    doc.font('Courier')
                        .fontSize(fontSizes.code)
                        .fillColor(colors.code)
                        .text(parts[i], xPos, yPos, { continued: true });
                }
            }
        }
        yPos += 14;
    }
    // Regular text
    else {
        // Handle bold
        let text = cleanLine;
        let isBold = false;
        
        if (text.includes('**')) {
            text = text.replace(/\*\*/g, '');
            isBold = true;
        }
        
        doc.font(isBold ? 'Helvetica-Bold' : 'Helvetica')
            .fontSize(fontSizes.body)
            .fillColor(colors.body)
            .text(text, 60, yPos, { width: 480 });
        
        // Calculate height based on text length
        const textHeight = Math.ceil(text.length / 80) * 14;
        yPos += textHeight + 4;
    }
}

// Finalize PDF
doc.end();

stream.on('finish', () => {
    console.log(`PDF gerado com sucesso: ${pdfFile}`);
});

stream.on('error', (err) => {
    console.error('Erro ao gerar PDF:', err);
});
